### PR TITLE
WWW-Authenticate header with any scheme #60

### PR DESCRIPTION
**Pull request by kkazuo**
I would like to use "Bearer" auth.

http://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml

### CODE REVIEW

File: lib/webmachine.ml [hunk 1]
```diff
@@ -103,9 +103,13 @@ module type S = sig
   type 'body provider = ('body, 'body) op
   type 'body acceptor = (bool, 'body) op
 
+  type www_authenticate =
+    { scheme : string; realm : string; params : (string * string) list }
+
   type auth =
     [ `Authorized
     | `Basic of string
+    | `Unauthorized of www_authenticate
```
Comment (seliopou) [inline lib/webmachine.ml]: `Challenge` is a more appropriate name for this variant, as it's the names used in [RFC7235](https://tools.ietf.org/html/rfc7235#section-4.1)

File: lib/webmachine.ml [hunk 2]
```diff
@@ -478,6 +486,12 @@ module Make(IO:IO) = struct
         | `Basic realm ->
           self#set_response_header "WWW-Authenticate" ("Basic realm=\"" ^ realm ^ "\"");
           self#halt 401
+        | `Unauthorized auth ->
+          self#set_response_header "WWW-Authenticate" (String.concat " " (
+              auth.scheme ::
+              ("realm=\"" ^ auth.realm ^ "\"") ::
+              (List.map (fun (k, v) -> k ^ "=\"" ^ v ^ "\"") auth.params)));
```
Comment (seliopou) [inline lib/webmachine.ml]: This code could benefit from the use of the [`Buffer`](https://caml.inria.fr/pub/docs/manual-ocaml/libref/Buffer.html) to minimize allocations.

**Comment by kkazuo** [comment]
I fixed the code.
Do you think?

Thanks

**Comment by seliopou** [comment]
Thanks for your contribution!

**Pull request merged**